oak-core/src/main/java/org/apache/jackrabbit/oak/query/QueryFormatter.java

```diff
@@ -0,0 +1,138 @@
+ */
+public class QueryFormatter {
+
+    /**
+     * Detect whether the query is an XPath query.
+     *
+     * @param query the query
+     * @param language the language, if known, or null
+     * @return true if xpath
+     */
+    public static boolean isXPath(String query, String language) {
+        if (language != null) {
+            return "xpath".equals(language);
+        }
+        query = query.trim().toLowerCase(Locale.ENGLISH);
+        // explain queries
+        if (query.startsWith("explain")) {
+            query = query.substring("explain".length()).trim();
+            if (query.startsWith("measure")) {
+                query = query.substring("explain".length()).trim();
+            }
+        }
+        // union queries
+        while (query.startsWith("(")) {
+            query = query.substring("(".length()).trim();
+        }
+        if (query.startsWith("select")) {
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Format the query into a more human-readable way, by adding newlines.
+     * Warning: newlines are also added inside e.g. string literals.
+     *
+     * @param query the query (may not be null)
+     * @param language the query language, or null if unknown
+     * @return the formatted query
+     */
+    public static String format(String query, String language) {
+        boolean xpath = isXPath(query, language);
+        if (xpath) {
+            return formatXPath(query);
+        } else {
+            return formatSQL(query);
+        }
+    }
+
+    private static String formatXPath(String query) {
+        StringBuilder buff = new StringBuilder(query);
+        for (int i = 0; i < buff.length(); i++) {
+            char c = buff.charAt(i);
+            if (c == '\'' || c == '"') {
+                while (++i < buff.length() && buff.charAt(i) != c) {
+                    // skip
+                }
+            } else if (c =='[') {
+                buff.insert(i + 1, "\n  ");
+                i += 3;
+            } else if (c <= ' ') {
+                String sub = buff.substring(i, Math.min(i + 10, buff.length()));
+                if (sub.startsWith(" and ")
+                        || sub.startsWith(" or ")
+                        || sub.startsWith(" order by ")
+                        || sub.startsWith(" option(")) {
+                    buff.setCharAt(i, '\n');
+                    buff.insert(i + 1, "  ");
+                    // just skip over the whitespace - but that's OK
+                    i += 2;
+                }
+            }
+        }
+        return buff.toString();
+    }
+
+    private static String formatSQL(String query) {
+        StringBuilder buff = new StringBuilder(query);
+        for (int i = 0; i < buff.length(); i++) {
+            char c = buff.charAt(i);
+            if (c == '\'' || c == '"') {
+                while (++i < buff.length() && buff.charAt(i) != c) {
+                    // skip
+                }
```

Comment:
Won't this fail if there is some string in the query that contains " inside it? For instance, a string like `"he said \"Hello\""`.

Reply:
Backslash is not an escape character. The escape character for ' is ', so that it would be:

    "he said ""Hello"""

and

    'he said ''Hello'''

Because of this, we just need to search for the next ' / ". The "" and '' are correctly recognised. There's also a test for that.